<?xml version="1.0"?>
<!-- Launch file for pipeline-->
<!-- Source: iros_method2.launch -->
<launch>
    <arg name="robot_name" default="sam"/>
    
    <!-- Node parameters -->
    <param name="robot_name" value="$(arg robot_name)"/>
    <param name="frame" value="odom"/>
    <param name="robot_frame" value="sam0"/>
    <param name="path_name" value="/home/sam/slam_ws/src/sam_slam/processing scripts/data/pipeline"/>

    <!-- Scenario parameters -->
    <param name="pipeline_scenario" value="True" />

    <!-- Data parameters -->
    <param name="dr_update_time" value="2.0"/>
    <param name="detect_update_time" value="0.5"/>

    <param name="camera_update_time" value="15.0"/>
    <param name="sss_update_time" value="15.0"/>
    <param name="sss_buffer_len" value="5"/>

    <!-- Detection and association parameters -->
    <param name="simulated_data" value="True"/>
    <param name="record_ground_truth" value="True"/>

    <param name="detector_seq_id_as_score" value="True"/>
    <param name="manual_associations" value="False"/>
    <param name="simulated_detections" value="True"/>
    <param name="rope_associations" value="True"/>
    <param name="prioritize_buoy_detections" value="True"/>
    <param name="da_distance_threshold" value="50"/>
    <param name="da_m_distance_threshold" value="-1.0"/>
    <param name="individual_rope_detections" value="True"/>
    <param name="update_priors" value="False"/>

    <param name="use_rope_detections" value="True"/>
    <param name="rope_batch_size" value="0"/>
    <param name="rope_batch_by_line" value="False"/>
    <param name="rope_batch_by_line_timeout" value="3"/>
    <param name="batch_by_swath" value="False"/>

    <!-- Detector data -->
    <!-- <param name= "buoy_detections_path" value="/home/julian/catkin_ws/src/sam_slam/processing scripts/data/image_process_buoys.csv"/> -->
    <!-- <param name= "rope_port_detections_path" value="/home/julian/catkin_ws/src/sam_slam/processing scripts/data/image_process_ropes_port.csv"/> -->
    <!-- <param name= "rope_star_detections_path" value="/home/julian/catkin_ws/src/sam_slam/processing scripts/data/image_process_ropes_star.csv"/> -->
    <arg name="detector_min_update_time" default="2.0" />

    <!-- Graph parameters -->
    <!-- Currently using default ISAM2 parameters -->
    <!-- Angles are in units of degrees, which are then converted to rads -->
    <!-- Prior sigmas -->
    <param name="prior_ang_sig_deg" value="10.0"/>  <!-- ICRA: 10.0 -->
    <param name="prior_dist_sig" value="1"/>  <!-- ICRA: 10.0  > 2.5-->
    <!-- Buoy -->
    <param name="buoy_dist_sig" value="1"/>
    <!-- Odometry -->
    <param name="odo_ang_sig_deg" value="0.5"/>   <!-- ICRA: 0.05 > 0.1 -->
    <param name="odo_dist_sig" value="0.1"/>   <!-- ICRA: 0.001 -->

    <!-- Detection sigmas -->
    <!-- buoys -->
    <param name="buoy_detect_ang_sig_deg" value="0.5"/>
    <param name="buoy_detect_dist_sig" value=".25"/>
    <!-- ropes -->
    <param name="detect_ang_sig_deg" value=".5"/>  <!-- ICRA: 25.0 > 0.5 -->
    <param name="detect_dist_sig" value="0.25"/>  <!-- ICRA: 15.0 -->

    <!-- Detection sigmas -->
    <param name="rope_along_sig" value="-0.5"/>  <!-- negative values will cause it to scale with rope length -->
    <param name="rope_cross_sig" value="1.0"/>

    <!-- Output parameters -->
    <param name="verbose_listener_DRs" value="false"/>
    <param name="verbose_listener_detections" value="True"/>
    <param name="verbose_listener_sonars" value="false"/>
    <param name="verbose_listener_buoys" value="true"/>
    <param name="verbose_listener_cameras" value="false"/>

    <param name="verbose_graph_update" value="True"/>
    <param name="verbose_graph_rope_detections" value="True"/>
    <param name="verbose_graph_buoy_detections" value="True"/>
    <param name="verbose_graph_rope_associations" value="True"/>
    <param name="verbose_graph_buoy_associations" value="False"/>

    <arg name="verbose_detector" default="True" />

    <!-- Map parameters -->
    <param name="pipeline_end_coords" value="[[-260, -829],
                                              [-263, -930],
                                              [-402, -1081],
                                              [-403, -1178]]" />

    <param name="pipeline_depth" value="-85" />

    <param name="pipeline_lines" value="[[0, 1],
                                         [1, 2],
                                         [2, 3]]" />

    <param name="pipeline_colors" value="[[1.0, 1.0, 0.0],
                                          [0.0, 1.0, 1.0],
                                          [1.0, 0.0, 1.0]]" />

    <!-- NODES -->
    <!-- start rviz -->
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find sam_slam)/rviz/rviz_pipeline.rviz"/>

    <!-- Sam SLAM -->
    <node name="sam_slam_listener" pkg="sam_slam" type="sam_listener_online_slam_node.py" output="screen"/>

    <!-- Play the ROS bag -->
    <node name="rosbag_player" pkg="rosbag" type="play" args="-r 1 --clock /home/sam/filtered.bag"/>

    <!-- scenario map and visualization -->
    <node name="pipeline_marker_node" pkg="sam_slam" type="pipeline_map_markers.py" output="screen" />

    <!-- detector -->
    <node name="detector_node" pkg="sam_slam" type="point_cloud_detector.py" output="screen">
        <param name="~detector_min_update_time" value="$(arg detector_min_update_time)" />
        <param name="~verbose_detector" value="$(arg verbose_detector)" />
    </node>

    <!-- dr and gt markers -->
    <node name="dr_gt_markers" pkg="sam_slam" type="pipeline_gt_dr_publisher_node.py" output="screen" />

</launch>
